orttraining/orttraining/python/training/ortmodule/_module_manager_interface.py

```diff
@@ -0,0 +1,121 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# _module_manager_interface.py
+
+from abc import ABC, abstractmethod
+import torch
+from typing import Iterator, Optional, Tuple, TypeVar, Set, Callable
+
+T = TypeVar('T', bound='Module')
+
+class ModuleManagerInterface:
+    """Abstract class that provides the function signatures for the torch.nn.Module
+
+    Concrete implementations should inherit from this class and provide necessary executions.
+    """
+    def __init__(self, module):
+        self._original_module = module
+
+    @abstractmethod
+    def forward(self):
+        """Executes the forward method for ORTModule
+
+        This is an abstract method and must be overridden by a concrete implementation.
+        """
+        pass
+
+    @abstractmethod
+    def _apply(self, fn):
+
+        pass
+
+    @abstractmethod
+    def apply(self: T, fn: Callable[['Module'], None]) -> T:
+
+        pass
+
+    @abstractmethod
+    def train(self: T, mode: bool = True) -> T:
+
+        pass
+
+    @abstractmethod
+    def state_dict(self, destination=None, prefix='', keep_vars=False):
+
+        pass
+
+    @abstractmethod
+    def load_state_dict(self, state_dict: 'OrderedDict[str, Tensor]',
+                        strict: bool = True):
+
+        pass
```

Comment:
```@abstractmethod``` decorator should raise an exception if the derived class does not have any implementation for the method.

Reply:
right, however lets look at onnx case. The OnnxModuleManager can give reasonable implementation to 3 functions out of 10ish. In order to prevent copy-pasted functions with the same code inside, that just raise exception, it will be nice to have default behavior in base class. The default behavior looks reasonable to raise unsupported exception.